spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/CommitTimestamp.java

```diff
@@ -0,0 +1,118 @@
+ * @since 1.3.0
+ */
+public final class CommitTimestamp {
+	private static final Map<Object, Object> VALUES = new ConcurrentHashMap<>();
+
+	static {
+		VALUES.put(com.google.cloud.Timestamp.class, Value.COMMIT_TIMESTAMP);
+	}
+
+	private CommitTimestamp() {
+	}
+
+	/**
+	 * Creates a value for {@link Value#COMMIT_TIMESTAMP}.
+	 * @param timestampClass a class to be presented as a Timestamp in the database column.
+	 * @param <T> type of the class. Supported by-default types are:
+	 * 	{@link com.google.cloud.Timestamp}, {@link java.sql.Timestamp}, {@link java.time.LocalDateTime},
+	 * 	{@link java.time.Instant}, {@link java.util.Date}.
+	 * @return a value that will be converted to {@link Value#COMMIT_TIMESTAMP} by {@link SpannerConverters}
+	 * @throws IllegalArgumentException when the {@code timestampClass} was not registered
+	 * 	 with the method {@link #register(CommitTimestampDecorator)}.
+	 * @see #register(CommitTimestampDecorator)
+	 */
+	@SuppressWarnings("unchecked")
+	public static <T> T of(Class<T> timestampClass) throws IllegalArgumentException {
+		return (T) VALUES.computeIfAbsent(timestampClass, key -> {
+			throw new IllegalArgumentException(
+					String.format("\"CommitTimestamp\" is not supported for the type %s", key));
+		});
+	}
+
+	/**
+	 * The method should be used when we need to register a custom "to timestamp" converter as {@link CommitTimestampDecorator}.
+	 * @param converter the {@link Converter} to register as CommitTimestampSupplier
+	 * @param <S> the "source" type of {@link Converter}
+	 * @param <C> the type if {@link CommitTimestampDecorator}
+	 * @return registered converter.
+	 * @throws IllegalStateException when the typed argument of the {@code converter} already registered.
+	 *   It is impossible to overwrite existing registrations because it could affect the core converters
+	 *   of the {@link SpannerConverters} class.
+	 */
+	@SuppressWarnings("unchecked")
+	public static <S, C extends CommitTimestampDecorator<S>> C register(C converter) throws IllegalStateException {
+		final Class<S> argument = (Class<S>) resolveTypeArgument(converter.getClass(), CommitTimestampDecorator.class);
+		VALUES.compute(argument, (key, old) -> {
+			if (old == null) {
+				return converter.commitTimestamp;
+			}
+			throw new IllegalStateException(
+					String.format("The value %s already registered as \"CommitTimestamp\" for the type %s", old, key));
+		});
+		return converter;
+	}
+
+	/**
+	 * A specific decorator of the "to timestamp" function that any custom converter should follow
+	 * to support "CommitTimestamp" feature.
+	 * @param <S> a source type of converter.
+	 */
+	public static abstract class CommitTimestampDecorator<S> implements Converter<S, Timestamp> {
+
+		final S commitTimestamp;
+		private final Function<S, Timestamp> converter;
+
+		protected CommitTimestampDecorator(S commitTimestamp, Function<S, Timestamp> converter) {
+			this.commitTimestamp = commitTimestamp;
+			this.converter = converter;
```

Comment:
is it possible call `CommitTimestamp.register(this)` here and avoid calling it from `SpannerComvertors` class?

Reply:
It was like you telling me from the very beginning.
Then I changed it because I thought you will not like when constructor doing something not tightly related to the construction of the object

Reply:
changing back...